src/main/knn/CheckHNSWConnectedness.java

```diff
@@ -0,0 +1,120 @@
+package knn;
+
+import org.apache.lucene.codecs.KnnVectorsReader;
+import org.apache.lucene.codecs.lucene95.Lucene95HnswVectorsReader;
+import org.apache.lucene.codecs.perfield.PerFieldKnnVectorsFormat;
+import org.apache.lucene.index.*;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.lucene.util.hnsw.HnswGraph;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.*;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+public class CheckHNSWConnectedness {
+    private static int getReachableNodes(HnswGraph graph, int level) throws IOException {
+        Set<Integer> visited = new HashSet<>();
+        Stack<Integer> candidates = new Stack<>();
+        candidates.push(graph.entryNode());
+
+        while (!candidates.isEmpty()) {
+            int node = candidates.pop();
+
+            if (visited.contains(node)) {
+                continue;
+            }
+
+            visited.add(node);
+            graph.seek(level, node);
+
+            int friendOrd;
+            while ((friendOrd = graph.nextNeighbor()) != NO_MORE_DOCS) {
+                candidates.push(friendOrd);
+            }
+        }
+        return visited.size();
+    }
+
+    private static int getOverallReachableNodes(HnswGraph graph, int numLevels) throws IOException {
+        Set<Integer> visited = new HashSet<>();
+        Stack<Integer> candidates = new Stack<>();
+        candidates.push(graph.entryNode());
+
+        while (!candidates.isEmpty()) {
+            int node = candidates.pop();
+
+            if (visited.contains(node)) {
+                continue;
+            }
+
+            visited.add(node);
+            // emulate collapsing all the edges
+            for (int level = 0; level < numLevels; level++) {
```

Comment:
for levels > 0 can we iterate over the nodes in the level rather than all the nodes?